Gems/EMotionFX/Code/EMotionFX/Tools/EMotionStudio/Plugins/StandardPlugins/Source/AnimGraph/ContextMenu.cpp

```diff
@@ -185,35 +153,22 @@ namespace EMStudio
 
                 if (actionFilter.m_createNodes)
                 {
-                    const AZStd::vector<EMotionFX::AnimGraphNode::ECategory> categories =
-                    {
-                        EMotionFX::AnimGraphNode::CATEGORY_SOURCES,
-                        EMotionFX::AnimGraphNode::CATEGORY_BLENDING,
-                        EMotionFX::AnimGraphNode::CATEGORY_CONTROLLERS,
-                        EMotionFX::AnimGraphNode::CATEGORY_PHYSICS,
-                        EMotionFX::AnimGraphNode::CATEGORY_LOGIC,
-                        EMotionFX::AnimGraphNode::CATEGORY_MATH,
-                        EMotionFX::AnimGraphNode::CATEGORY_MISC
-                    };
-
                     // Create nodes in palette view for each of the categories that are possible to be added to the currently focused graph.
                     EMotionFX::AnimGraphNode* currentNode = nodeGraph->GetModelIndex().data(AnimGraphModel::ROLE_NODE_POINTER).value<EMotionFX::AnimGraphNode*>();
 
                     auto* action = new QWidgetAction(menu);
-                    auto* rootItem = new GraphCanvas::NodePaletteTreeItem("root", AnimGraphEditorId);
+
+                    NodePaletteModelHelper modelHelper{ plugin };
+                    modelHelper.InitForNode(currentNode);
+
                     GraphCanvas::NodePaletteConfig config;
-                    config.m_rootTreeItem = rootItem;
+                    config.m_rootTreeItem = modelHelper.GetRootItem();
```

Comment:
question: Is there a potential lifetime issue here when `modelHelper` goes out of scope as we're taking a reference to `modelHelper.GetRootItem()` below and it is a local variable? It might be fine but I just wanted to double check

Reply:
I don't thing there's a lifetime issue since the palette widget takes ownership of the root pointer. It's passed in the `config` to the widget in `m_palette->SetupNodePalette(config);` couple of lines below.

Inside `SetupNodePalette` the root item is passed to the `GraphCanvasTreeModel` which in turns takes ownership of the item by storing it in a unique pointer:

```cpp
    class GraphCanvasTreeModel
        : public QAbstractItemModel
    {
/*snip*/
        AZStd::unique_ptr<GraphCanvasTreeItem> m_treeRoot;
    };
```

The only thing for the user of `NodePaletteModelHelper` is to not use it after the palette widget is destroyed because the item pointer will be invalid.

Since on one hand, the palette widget takes ownership of the root node and on the other, we still need a reference to it when updating the tree contents I don't see an easy way out. I could devise some class that would tie the palette widget and the helper class but that might be an overkill.

This `NodePaletteModelHelper` is more of a builder one might say. It can however rebuild the tree when it gets reinitialized for different focused node.  Maybe `NodePaletteModelUpdater` would make more sense?

Reply:
I went ahead and renamed the class to `NodePaletteModelUpdater`. It's a bit more descriptive than `*Helper`

Reply:
Okay I see, thanks for the explanation @MiKom. The updated name definitely helps 👍